### DISCUSSION TITLE
[A-L] (12 gennaio 2023) Esercizio 3

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#73] [Esami A-L]
_tengo a precisare che sulla traccia originale ha scritto 2016, si è sbagliato, è del 2023._
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/4bb60e2c-ce0b-4d00-91cd-a83dd13a0600)

**Comment by Frisbon** [comment]
### Esempio (per chi non capisce la traccia)
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/8196d4ff-a8ab-404a-b48b-9238c6a03425)

**Comment by Frisbon** [comment]
```python
def DFS_ric(VIS,x,target,cr,cb,result):
    VIS[x] = 1;
    for y in x.out:
        
        if y == target:
            if y == "RED": cr+=1
            else: cb+=1
        
            if cr == cb: return True
            else: return False
        
        if VIS[y] == 0:
            if y == "RED": result = result or DFS_ric(VIS,y,target,cr+1,cb)
            else: result = result or DFS_ric(VIS,y,target,cr,cb+1)
    
    return result
    
    
def es3(G,x,y):
    VIS = [0,...,0]; VIS[x] = 1
    cr,cb = 0
    if x == "RED": cr = 1
    else: cb = 1
    return DFS_ric(VIS,x,y,cr,cb,False)
```
Valutazione di GPT (non so per il costo)
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/ef4ad78f-e1b7-4152-870d-556c76dfbbb1)

**Reply by faciolox** [reply]
Qua uso una DFS per trovare tutti i cammini e nel frattempo mi salvo anche i numeri degli archi blu e rossi. Quando trovi un cammino se blu = rossi allora lo salvo in un insieme

<img width="1009" alt="Screenshot 2024-06-17 alle 15 34 05" src="https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/99180436/a5fc8544-ce73-4df4-8905-30693853ad51">

**Comment by niccolo-fato** [comment]


**Reply by salicius** [reply]
Perdona l'ignoranza, cosa intendi per "u" all'interno della dfs? Il professore l'ha ritenuto corretta questa esecuzione?

**Comment by CiottoloMaggico** [comment]
Soluzione che utilizza tabella costruita dinamicamente
![immagine](https://github.com/user-attachments/assets/9a1d5573-9369-4ec3-8bb9-039c0bc84a3a)

**Comment by pippolippoculonero** [comment]
![image](https://github.com/user-attachments/assets/c96a29cc-1ccf-43b7-9de5-c0483524f593)
non sono convinto della correttezza del codice, dovrebbe essere in costo O(n+m)

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/b6d226be-4705-49ea-9c7c-ace417c3ecf0)

**Comment by cek753** [comment]
Utilizzo un array A di supporto in modo da tenere traccia delle diverse coppie (r,b) che arrivano ad un elemento dell’array partendo da x 

<img width="1388" height="1148" alt="image" src="https://github.com/user-attachments/assets/a5e50397-e940-46d8-9491-0808576e4a14" />
